webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/annotation/TSAnnotationService.scala

```diff
@@ -573,6 +574,27 @@ class TSAnnotationService @Inject()(val remoteWebknossosClient: TSRemoteWebknoss
     } yield ()
   }
 
+  private def flushSkeletonTracingWithSeparatedTreeBodies(
+      annotationWithTracings: AnnotationWithTracings,
+      skeletonTracingId: String,
+      skeletonTracing: SkeletonTracing)(implicit ec: ExecutionContext): Fox[Unit] =
+    for {
+      // All tree bodies that were changed by the update actions need to be stored
+      updatedTreeIds <- annotationWithTracings.getUpdatedTreeBodyIdsForSkeleton(skeletonTracingId).toFox
+      // All tree bodies that were previously not stored externally need to be stored
+      treeBodyIdsToMigrate = skeletonTracing.trees.filter(!_.getHasExternalTreeBody).map(_.treeId).toSet
+      nowPresentTrees = skeletonTracing.trees.map(_.treeId).toSet
+      treeIdsToFlush = nowPresentTrees.intersect(updatedTreeIds.union(treeBodyIdsToMigrate))
```

Comment:
Why is the `nowPresentTrees` helpful? I cant see a case where `treeBodyIdsToMigrate` contains more treeIds than `nowPresentTrees`.

- `treeBodyIdsToMigrate` is derived from `skeletonTracing.trees` then filtered and then mapped to their ids.
- `nowPresentTrees` is also derived from `skeletonTracing.trees` unfiltered and then mapped to their ids.

So to me `nowPresentTrees` always contains all treeIds of `treeBodyIdsToMigrate` and thus the intersection should always yield `treeBodyIdsToMigrate`.

Or what am I missing here :thinking:

Reply:
It is updatedTreeIds that may contain more than the now present ones: it’s deleted ones. Those are strictly speaking updated (and the updateActions report so), but cannot be flushed, as they are no longer present.
Alternatively, the update action could skip the deleted ones when it reports the updated ids, but that doesn’t feel very clean. What do you think?